src/internal/observable/defer.ts

```diff
@@ -52,7 +52,9 @@ import { EMPTY } from './empty';
  * @name defer
  * @owner Observable
  */
-export function defer<R extends ObservableInput<any> | void>(observableFactory: () => R): Observable<ObservedValueOf<R>> {
+export function defer<R extends ObservableInput<any> | void>(
+  observableFactory: [R] extends [undefined] | [void] ? never : () => R
+): Observable<ObservedValueOf<R>> {
```

Comment:
IMO, if we are going to fix this footgun, we should only be accepting factories that return a valid `ObservableInput`. I think the fix should be the removal of the `void`:

```ts
export function defer<R extends ObservableInput<any>>(observableFactory: () => R): Observable<ObservedValueOf<R>> { ...
```

And I think the implementation should be changed, too. If it's not, the footgun is still there for non-TS usage.

Reply:
I agree. It's weird that we ever had this work.

Reply:
Done, I've updated the tests and dstlint spec to reflect the changes